onnxruntime/core/providers/cpu/nn/layer_norm_impl.cc

```diff
@@ -96,58 +101,61 @@ void ComputeJob(
     U* mean_data,
     U* inv_std_dev_data,
     AllocatorPtr alloc) {
-  ORT_UNUSED_PARAMETER(scale_data);  // only used in float/double overload
-  ORT_UNUSED_PARAMETER(bias_data);   // only used in float/double overload
-
-  const MLFloat16* p_input = X_data + task_idx * norm_size;
-  MLFloat16* p_output = Y_data + task_idx * norm_size;
+  ORT_UNUSED_PARAMETER(scale_data);
+  ORT_UNUSED_PARAMETER(bias_data);
+  ORT_UNUSED_PARAMETER(alloc);
 
-  float mean(0.0f);
-  float mean_square(0.0f);
+  // reinterpret input/output MLFloat16* as Eigen::half*
+  const Eigen::half* p_input = reinterpret_cast<const Eigen::half*>(X_data + task_idx * norm_size);
+  Eigen::half* p_output = reinterpret_cast<Eigen::half*>(Y_data + task_idx * norm_size);
 
-  const size_t num_elems = static_cast<size_t>(norm_size);
-  IAllocatorUniquePtr<float> input_float_uptr = IAllocator::MakeUniquePtr<float>(alloc, num_elems);
-  MlasConvertHalfToFloatBuffer(p_input, input_float_uptr.get(), num_elems);
+  // Fix: cast norm_size to Eigen::Index
+  Eigen::Map<const Eigen::Matrix<Eigen::half, Eigen::Dynamic, 1>> input_vec(
+      p_input, ToEigenIndex(norm_size));
+  Eigen::Map<Eigen::Matrix<Eigen::half, Eigen::Dynamic, 1>> output_vec(
+      p_output, ToEigenIndex(norm_size));
 
-  IAllocatorUniquePtr<float> output_float_uptr = IAllocator::MakeUniquePtr<float>(alloc, num_elems);
-  float* output_float_ptr = output_float_uptr.get();
+  // Compute mean and mean_square in float for precision
+  float mean = 0.0f;
+  float mean_square = 0.0f;
 
-  const float* input_float_ptr = input_float_uptr.get();
-  for (size_t h = 0; h < num_elems; h++) {
-    output_float_ptr[h] = input_float_ptr[h];
-    mean += input_float_ptr[h];
-    mean_square += input_float_ptr[h] * input_float_ptr[h];
+  for (int64_t i = 0; i < norm_size; ++i) {
+    float val = gsl::narrow_cast<float>(input_vec[ToEigenIndex(i)]);
```

Comment:
general nit: it's not a narrowing conversion from `Eigen::half` to `float`. maybe `static_cast` would be clearer.